### PR TITLE
Stacklevel argument updated #46687

### LINKED ISSUE #46687: CLN find stacklevel inside deprecate_nonkeyword_arguments
xref https://github.com/pandas-dev/pandas/pull/44896/files#r845356050

Rather than passing stacklevel to the decorator, e.g.

https://github.com/pandas-dev/pandas/blob/e1b95fa128203e6e21d3a1a6469c94e49f503e0c/pandas/io/parsers/readers.py#L831-L833

`find_stacklevel` should probably be called within `deprecate_nonkeyword_arguments`

### PR DESCRIPTION
**Pull request by Anupam-USP**
- [x] closes #46687 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by Anupam-USP** [comment]
Can you suggest me some test to write for these?

File: pandas/util/_decorators.py [hunk 1]
```diff
@@ -260,7 +261,7 @@ def future_version_msg(version: str | None) -> str:
 def deprecate_nonkeyword_arguments(
     version: str | None,
     allowed_args: list[str] | None = None,
-    stacklevel: int = 2,
+    stacklevel: int = None,
```
Comment (MarcoGorelli) [inline pandas/util/_decorators.py:264]: if you're defining it below (L308), you don't need it as argument any more, do you?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
this is just an internal refactor, so as long as the current test-suite passes, no new tests should be necessary

File: pandas/util/_decorators.py (hunk 1 above)
Comment (Anupam-USP) [reply pandas/util/_decorators.py:264]: No, I will remove it. Also should I check other files for definition of stacklevel and change them?

Comment (MarcoGorelli) [reply pandas/util/_decorators.py]: you should remove it from the docstring as well

**Comment by MarcoGorelli** [comment]
currently failing with
```
ImportError while loading conftest 'D:\a\pandas\pandas\pandas\conftest.py'.
pandas\__init__.py:48: in <module>
    from pandas.core.api import (
pandas\core\api.py:48: in <module>
    from pandas.core.groupby import (
pandas\core\groupby\__init__.py:1: in <module>
    from pandas.core.groupby.generic import (
pandas\core\groupby\generic.py:73: in <module>
    from pandas.core.frame import DataFrame
pandas\core\frame.py:202: in <module>
    from pandas.core.series import Series
pandas\core\series.py:5865: in <module>
    Series._add_numeric_operations()
pandas\core\generic.py:10950: in _add_numeric_operations
    @deprecate_nonkeyword_arguments(
E   TypeError: deprecate_nonkeyword_arguments() got an unexpected keyword argument 'stacklevel'
```

**Comment by Anupam-USP** [comment]
Hey @MarcoGorelli I tried changing the files and ran conftest.py to check, but getting errors like `AttributeError: module 'pandas._testing' has no attribute 'make_rand_series'` so not able to check the changes I have made. What can I do?

**Comment by MarcoGorelli** [comment]
hey @Anupam-USP - what command did you run that gave you that output?

You might need to re-compile the C extensions (`python setup.py build_ext -j 4`)

**Comment by Anupam-USP** [comment]
Only simple python command to run conftest.py file

**Comment by Anupam-USP** [comment]
To overcome the aforementioned issue, I have to still declare it as I did previously.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
To overcome
```
E   TypeError: deprecate_nonkeyword_arguments() got an unexpected keyword argument 'stacklevel'
```
you should remove calls of `deprecate_nonkeyword_arguments` where `stacklevel` is passed

To address the other issue, you probably just need to recompile the C extensions (`python setup.py build_ext -j 4`). There should be no need to keep `stacklevel` as an argument to `deprecate_nonkeyword_arguments`

File: pandas/util/_decorators.py [hunk 2]
```diff
@@ -304,6 +305,7 @@ def decorate(func):
             f"{future_version_msg(version)} all arguments of "
             f"{name or func.__qualname__}{{arguments}} will be keyword-only."
         )
+        stacklevel = find_stack_level()
```
Comment (MarcoGorelli) [inline pandas/util/_decorators.py:308]: in L317, you should just set
```
stacklevel=find_stack_level()
```
rather than assigning it here

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@Anupam-USP what's new entry not needed as this is not public (i removed it)

**Comment by MarcoGorelli** [comment]
As noted here https://github.com/pandas-dev/pandas/issues/47385 , I think the test failure might be a false-positive. Will try to address that one first

**Review by MarcoGorelli — APPROVED** [review]
Have fixed this up, I think it should be fine now and that the failed job had an unrelated failure (re-running to check)

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @Anupam-USP and @MarcoGorelli

**Pull request merged**